3rdparty/openvx/include/ivx.hpp

```diff
@@ -0,0 +1,1611 @@
+            else throw WrapperError(std::string(__func__)+"(): wrong plane index");
+            break;
+        case VX_DF_IMAGE_NV12:
+        case VX_DF_IMAGE_NV21:
+            if (planeIdx == 0u) m.create(h, w, formatToMatType(f, 0));
+            else if (planeIdx == 1u)  m.create(h/2, w/2, formatToMatType(f, 1));
+            else throw WrapperError(std::string(__func__)+"(): wrong plane index");
+            break;
+        case VX_DF_IMAGE_RGB:
+        case VX_DF_IMAGE_RGBX:
+        case VX_DF_IMAGE_UYVY:
+        case VX_DF_IMAGE_YUYV:
+        case VX_DF_IMAGE_U8:
+        case VX_DF_IMAGE_U16:
+        case VX_DF_IMAGE_S16:
+        case VX_DF_IMAGE_U32:
+        case VX_DF_IMAGE_S32:
+        case /*VX_DF_IMAGE_F32*/VX_DF_IMAGE('F', '0', '3', '2'):
+            if(planeIdx == 0u) m.create(h, w, formatToMatType(f));
+            else throw WrapperError(std::string(__func__)+"(): wrong plane index");
+            break;
+        default: throw WrapperError(std::string(__func__)+"(): unsupported color format");
+        }
+    }
+
+    void copyTo(vx_uint32 planeIdx, cv::Mat& m)
+    {
+        createMatForPlane(m, planeIdx);
+        copyTo(planeIdx, createAddressing((vx_uint32)m.cols, (vx_uint32)m.rows, (vx_int32)m.elemSize(), (vx_int32)m.step), m.ptr());
+    }
+
+    void copyFrom(vx_uint32 planeIdx, const cv::Mat& m)
+    {
+        // TODO: add sizes consistency checks
+        //vx_rectangle_t r = getValidRegion();
+        copyFrom(planeIdx, createAddressing((vx_uint32)m.cols, (vx_uint32)m.rows, (vx_int32)m.elemSize(), (vx_int32)m.step), m.ptr());
+    }
+#endif //IVX_USE_OPENCV
+
+    struct Patch;
+};
+
+struct Image::Patch
+{
+public:
+    const vx_imagepatch_addressing_t& addr() const
+    { return _addr;}
+
+    void* data() const
+    { return _data; }
+
+#ifdef VX_VERSION_1_1
+    vx_memory_type_e memType() const
+    { return _memType; }
+
+    vx_map_id mapId() const
+    { return _mapId; }
+#else
+    const vx_rectangle_t& rect() const
+    { return _rect; }
+
+    vx_uint32 planeIdx() const
+    { return _planeIdx; }
+#endif // VX_VERSION_1_1
+
+    vx_image img() const
+    { return _img; }
+
+    bool isMapped() const
+    { return _img != 0; }
+
+    bool isMemoryExternal() const
+    { return _externalMem; }
+
+#ifdef IVX_USE_OPENCV
+    cv::Mat& getMat()
+    {
+        if(_m.empty() && _data)
+        {
+            int matType = CV_8UC(_addr.stride_x);
+            if(_img)
+            {
+                vx_df_image format;
+                IVX_CHECK_STATUS( vxQueryImage(_img, VX_IMAGE_FORMAT, &format, sizeof(format)) );
+                matType = formatToMatType(format);
+            }
+            _m = cv::Mat( vx_int32((vx_int64)_addr.dim_y * VX_SCALE_UNITY / _addr.scale_y),
+                          vx_int32((vx_int64)_addr.dim_x * VX_SCALE_UNITY / _addr.scale_x),
+                          matType, _data, std::size_t(_addr.stride_y) );
```

Comment:
This Mat doesn't control lifetime of underlying data.

It is better to replace implementations of alloc/free in case of "Copy" access:
- allocate Mat directly: with passed information about datatype (use CV_8UCn in case of unsupported types) and size
- use `_m.data` instead of `_data`

This class has over-complicated design. My suggestion is to split map/unmap (scoped access) and copyTo/copyFrom modes to restrict use cases (and make them like a "constructors").
These use cases seems buggy in the current implementation:
- sequence of `copyFrom` calls with different images. Memory is not allocated again, it is just reused.
- `map` for small image + `copyFrom` for larger image. Same problem.